ledger/ledger-runner-common/src/main/scala/com/daml/ledger/runner/common/CliConfig.scala

```diff
@@ -111,607 +112,643 @@ object CliConfig {
-            val maxContractKeyStateCacheSize = kv
-              .get("contract-key-state-cache-max-size")
-              .map(_.toLong)
-              .getOrElse(IndexServiceConfig.DefaultMaxContractKeyStateCacheSize)
-            val partConfig = CliParticipantConfig(
-              mode = runMode,
-              participantId = participantId,
-              address = address,
-              port = port,
-              portFile = portFile,
-              serverJdbcUrl = jdbcUrl,
-              indexerConfig = IndexerConfig(
-                startupMode = IndexerStartupMode.MigrateAndStart(allowExistingSchema = false),
-                maxInputBufferSize = indexerMaxInputBufferSize,
-                inputMappingParallelism = indexerInputMappingParallelism,
-                batchingParallelism = indexerBatchingParallelism,
-                ingestionParallelism = indexerIngestionParallelism,
-                submissionBatchSize = indexerSubmissionBatchSize,
-                enableCompression = indexerEnableCompression,
-              ),
-              apiServerDatabaseConnectionPoolSize = apiServerConnectionPoolSize,
-              apiServerDatabaseConnectionTimeout = apiServerConnectionTimeout,
-              managementServiceTimeout = managementServiceTimeout,
-              maxContractStateCacheSize = maxContractStateCacheSize,
-              maxContractKeyStateCacheSize = maxContractKeyStateCacheSize,
-            )
-            config.copy(participants = config.participants :+ partConfig)
-          })
+        ),
+      builder.checkConfig(checkFileCanBeRead),
+    )
+  }
 
-        opt[String]("ledger-id")
-          .optional()
-          .text(
-            "The ID of the ledger. This must be the same each time the ledger is started. Defaults to a random UUID."
-          )
-          .action((ledgerId, config) => config.copy(ledgerId = ledgerId))
+  private def commandRunLegacy[Extra](
+      getEnvVar: String => Option[String],
+      extraOptions: OParser[_, CliConfig[Extra]],
+  ): OParser[_, CliConfig[Extra]] =
+    OParser
+      .builder[CliConfig[Extra]]
+      .cmd("run-legacy")
+      .text(
+        "Run Sandbox-On-X in a legacy mode with cli-driven configuration."
+      )
+      .action((_, config) => config.copy(mode = Mode.RunLegacy))
+      .children(legacyCommand(getEnvVar, extraOptions))
 
-        opt[String]("pem")
-          .optional()
-          .text(
-            "TLS: The pem file to be used as the private key. Use '.enc' filename suffix if the pem file is encrypted."
-          )
-          .action((path, config) =>
-            config.withTlsConfig(c => c.copy(keyFile = Some(new File(path))))
-          )
+  private def commandConvertConfig[Extra](
+      getEnvVar: String => Option[String],
+      extraOptions: OParser[_, CliConfig[Extra]],
+  ): OParser[_, CliConfig[Extra]] =
+    OParser
+      .builder[CliConfig[Extra]]
+      .cmd("convert-config")
```

Comment:
This option works without passing any CLI config and outputs the default values, as I see. I have two observations:
1. It does not print a default participant config (e.g. participants is empty `participants {}`). 
2. I am wondering whether it really helps that much to list everything: If the user does not know which values are defaults, he also can't easily re-use the output configuration in future deployments (e.g. what if defaults change and he accidentally overrides it with an older generated config file)>

Reply:
`convert-config` accepts same parameters as current not-migrated-to-hocon CLI. So if you run this mode with something like
`convert-config --participant participant-id=example,port=6865,contract-state-cache-max-size=2,contract-key-state-cache-max-size=2` it will generate a file for you.

I thought about the idea of having only overrides, but it does not sound like it is easy :)

Reply:
> I thought about the idea of having only overrides, but it does not sound like it is easy :)

Probably special logic is needed for deriving the explicitly set CLI params for creating an override. Probably not worth the effort